zookeeper-server/src/main/java/org/apache/zookeeper/server/SyncRequestProcessor.java

```diff
@@ -178,27 +226,22 @@ public void run() {
                 ServerMetrics.getMetrics().SYNC_PROCESSOR_QUEUE_TIME.add(startProcessTime - si.syncQueueStartTime);
 
                 // track the number of records written to the log
-                if (!si.isThrottled() && zks.getZKDatabase().append(si)) {
-                    if (shouldSnapshot()) {
-                        resetSnapshotStats();
+                if (!si.isThrottled() && zkDB.append(si)) {
+                    if (selfSnapTxnRollThreshold.meet(zkDB.getTxnCount(), zkDB.getTxnSize())) {
                         // roll the log
-                        zks.getZKDatabase().rollLog();
-                        // take a snapshot
-                        if (!snapThreadMutex.tryAcquire()) {
-                            LOG.warn("Too busy to snap, skipping");
+                        zkDB.rollLog();
+
+                        if (onlySnapWhenSafetyIsThreatened) {
+                            if (safetySnapThreshold.meet(zkDB.getTxnsSinceLastSnap(), zkDB.getTxnsSizeSinceLastSnap())) {
+                                snapGenerator.takeSnapshot(false);
```

Comment:
should we pass snapGenerator.fsyncSnapshotFromScheduler as parameter here instead of hardcoding a false?

Reply:
also if we want to hard code a value, it seems true is more safe than false here for durability guarantees - but i lose track of what our default options were when taking snapshot (fsync it or not, especially if IIRC we lost the fsync parameter when introducing SnapStream..)